pkg/controller/upgrade_funcs.go

```diff
@@ -879,3 +881,280 @@ func InstanceKeysRegionScopedName(ctx context.Context, objStore objstore.KVStore
 
 	return nil
 }
+
+type CloudletPoolKey struct {
+	// Name of the organization this pool belongs to
+	Organization string `protobuf:"bytes,1,opt,name=organization,proto3" json:"organization,omitempty"`
+	// CloudletPool Name
+	Name string `protobuf:"bytes,2,opt,name=name,proto3" json:"name,omitempty"`
+}
+
+type CloudletPool struct {
+	// Fields are used for the Update API to specify which fields to apply
+	Fields []string `protobuf:"bytes,1,rep,name=fields,proto3" json:"fields,omitempty"`
+	// CloudletPool key
+	Key CloudletPoolKey `protobuf:"bytes,2,opt,name=key,proto3" json:"key"`
+	// Cloudlets part of the pool
+	Cloudlets []edgeproto.CloudletKey `protobuf:"bytes,3,rep,name=cloudlets,proto3" json:"cloudlets"`
+	// Created at time
+	CreatedAt distributed_match_engine.Timestamp `protobuf:"bytes,4,opt,name=created_at,json=createdAt,proto3" json:"created_at"`
+	// Updated at time
+	UpdatedAt distributed_match_engine.Timestamp `protobuf:"bytes,5,opt,name=updated_at,json=updatedAt,proto3" json:"updated_at"`
+	// Preparing to be deleted
+	DeletePrepare bool `protobuf:"varint,6,opt,name=delete_prepare,json=deletePrepare,proto3" json:"delete_prepare,omitempty"`
+}
+
+type AutoProvCloudlet struct {
+	// Cloudlet key
+	Key edgeproto.CloudletKey `protobuf:"bytes,1,opt,name=key,proto3" json:"key"`
+	// Cloudlet location
+	Loc distributed_match_engine.Loc `protobuf:"bytes,2,opt,name=loc,proto3" json:"loc"`
+}
+
+type TrustPolicyExceptionKey struct {
+	// App Key
+	AppKey edgeproto.AppKey `protobuf:"bytes,1,opt,name=app_key,json=appKey,proto3" json:"app_key"`
+	// CloudletPool Key
+	CloudletPoolKey CloudletPoolKey `protobuf:"bytes,2,opt,name=cloudlet_pool_key,json=cloudletPoolKey,proto3" json:"cloudlet_pool_key"`
+	// TrustPolicyExceptionKey name
+	Name string `protobuf:"bytes,3,opt,name=name,proto3" json:"name,omitempty"`
+}
+
+// ZoneFeature adds support for a Zone feature. Each pre-existing cloudlet
+// will get a Zone of the same name created it for, and assigned to it.
+// Existing cloudlet pools will be rewritten as zone pools, with the same
+// membership (because on upgrade, cloudlets and zones are 1:1).
+func ZoneFeature(ctx context.Context, objStore objstore.KVStore, allApis *AllApis, sup *UpgradeSupport, dbModelID int32) error {
+	log.SpanLog(ctx, log.DebugLevelUpgrade, "Zone Feature")
+
+	// Create a new Zone for every Cloudlet and assign the cloudlet to the zone.
+	cloudletKeys, err := getDbObjectKeys(objStore, "Cloudlet")
+	if err != nil {
+		return err
+	}
+	for cloudletKey := range cloudletKeys {
+		_, err := objStore.ApplySTM(ctx, func(stm concurrency.STM) error {
+			cloudletStr := stm.Get(cloudletKey)
+			if cloudletStr == "" {
+				// deleted in the meantime
+				return nil
+			}
+			cloudlet := edgeproto.Cloudlet{}
+			if err2 := unmarshalUpgradeObj(ctx, cloudletStr, &cloudlet); err2 != nil {
+				return err2
+			}
+			if cloudlet.DbModelId >= dbModelID {
+				// already upgraded
+				return nil
+			}
```

Comment:
Are you thinking we might hit an issue where we'd need to run this upgrade function multiple times?